File: 0110-balanced-binary-tree/0110-balanced-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isBalanced = function(root) {
  const isBalance = Math.abs(getHeight(root?.left) - getHeight(root?.right)) <= 1;

  if (root?.left) {
    if (!isBalanced(root.left)) {
      return false;
    };
  }

  if (root?.right) {
    if (!isBalanced(root.right)) {
      return false;
    };
  }

  return isBalance;
  
  function getHeight(node) {
    if (!node) {
      return 0;
    }

    return Math.max(getHeight(node.left) + 1, getHeight(node.right) + 1)
  }
};

File: 0110-balanced-binary-tree/0110-balanced-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        if root is None:
            return True
        
        return self.get_height(root) != None

    def get_height(self, root: Optional[TreeNode]) -> bool or int:
        if root is None:
            return 0
        
        left_height = self.get_height(root.left)
        right_height = self.get_height(root.right)
        
        if right_height != None and left_height != None and -1 <= (left_height - right_height) <= 1:
            return max(left_height, right_height) + 1
        else:
            return None
          